### ISSUE TITLE
Callback-based variable queries

### ISSUE THREAD

**Issue by MOARdV** [Mihara/RasterPropMonitor#343]
For the modules that query numeric values, would a callback method work better?  Example:

JSIVariableAnimator calls comp.ProcessVariable potentially 3 times per Update call, per variable (assuming worst-case of using a named variable for each end of the range it's testing).  Each ProcessVariable call involves acquiring comp (dictionary search) and querying the variable (either computing it or doing another dictionary search).  Some variables are queried by many sources.  These searches add up computationally, and for some variables that don't update frequently (for example, CUSTOM_ALCOR_BACKLIGHT_ON), that's a lot of wasted cycles polling a rarely-changing value.  On the other hand, many of those props use a lower refreshRate (ignoring a number of update cycles before refreshing values).

I imagine this is less valuable in an all-glass cockpit, since I don't see the callback system being usable on a monitor display (multiple variables per line).  It increases the memory footprint, since I'd have a list of callbacks per registered variable.  Prop updates would move to FixedUpdate (since that's when it makes sense, after I clear and recompute variables).  I may experiment with this to see if I can find a way to quantify its impact.

**Comment by MOARdV** [comment]
I've made a couple of stabs at this, and I even managed to make a partial step in the right direction by reducing the number of VariableOrNumber objects that are created (consider that each variable animator creates 3 of them, and there can be multiple animators per prop...), but ... the issue always refers back to the vessel changing (due to undocking).  Although, now that I type this, I think I see a way to get around that problem, at least, in part.  I think, ultimately, that I will need to disallow / discourage the use of per-prop persistent vars.  They're a nuisance, and they're of limited use, anyway.

**Comment by MOARdV** [comment]
I'm punting on the callback system for the foreseeable future - I was able to get something put together, but it didn't work right for some custom variables.  I've kept the optimization of creating only one VariableOrNumber per variable name (and sharing that between all consumers), which reduces the memory footprint - in my current test platform, I had about 280 VariableOrNumber objects, versus well over a thousand before the optimization.

**Comment by MOARdV** [comment]
Reopening - with the level of complexity of some of the props and IVAs, I think callbacks for some variables may be extremely helpful (things such as backlights, status indicators for RCS, BRAKES, etc).

**Comment by MOARdV** [comment]
Reclosing - I've got code in JSIVariableAnimator that exits early if the scaled variable hasn't by at least a small epsilon.  This bypasses the processing that could cause performance lag (updating transforms, materials, etc).  A callback system won't improve things if we still have to call Update / FixedUpdate all the time.

**Comment by MOARdV** [comment]
Once again, I'm going to see if I can find a way for this to work.

**Comment by MOARdV** [comment]
It looks like it's coming together so far - it's a limited variant of JSIVariableAnimator, but it has the upside of very high response rate for those low-frequency values, such as backlights and switch position animators.

**Comment by MOARdV** [comment]
There is a definite benefit to this change: with a commonly used variable, such as `CUSTOM_ALCOR_BACKLIGHT_ON`, used for backlights on many of the ASET props, we go from one or two queries _per prop_ every 10 Updates to one query per FixedUpdate, and we eliminate all of the per-prop Update calls that increment timers.  The downside is there are a lot of prop configs to update in the ASET package.  I've got most of the prop configs done in my own package since I'm using them for development.

Since the callback code was copied from JSIVariableAnimator, it also exposed what appears to be a bug in JSIVariableAnimator: for the threshold behaviors (TurnOn and TurnOff), it looks like the various animations were wired backwards (scaleStart for On, scaleEnd for Off).  Fixing that in JSIVariableAnimator may break props, but I'll wire them correctly in JSICallbackAnimator.

**Comment by MOARdV** [comment]
As part of this change, I added a "profiler" that tracks how many queries are made for each variable, and how many FixedUpdate() calls are made.  When I initially tried it out with my Yarbrough Mk. 1-1 A2 IVA, I saw `CUSTOM_ALCOR_BACKLIGHT_ON` getting queried over 14 times per FixedUpdate.  Not every variable is queried every update - most props wait 5 or 10 updates (as opposed to FixedUpdate) to make their query, which means there are periodic surges of queries.

By adding the callbacks, I got that number down to about 9 per FixedUpdate - the remainder came from VariableLabel queries being used to control backlight effects.

**Comment by MOARdV** [comment]
Some test cases need checked out (eg - docking / undocking).  I most likely need to do the following:

When the RPMVC OnDestroy fires, call all of the callbacks one final time to provide a "I'm leaving" notification that can be used to unsubscribe / subscribe.  This assumes that the disappearing the prop's vessel field has already been updated to the new craft.

Add a "Does this vessel have a computer" query, since RPMVC.Instance generates an error if the vessel is not found.  That way, the callbacks can determine if their current vessel has an RPMVC before trying to subscribe (I found that RPMVC's OnDestroy executed before the props' on OnDestroy - so props couldn't unsubscribe if they were on a craft getting torn down).

See if there's a callback for "on vessel created" or "on part changes ownership" since the one hole I see in this design is on undock or staging.  If the player's craft is a newly-created craft, then the props won't know they are new, unless KSP destroys the old parts and creates new parts instead of changing the owners of the parts.
